interfaces/builtin/fwupd.go

```diff
@@ -0,0 +1,229 @@
+      member={Request,Release}Name
+      peer=(name=org.freedesktop.DBus),
+
+  dbus (receive, send)
+      bus=system
+      path=/org/freedesktop/DBus
+      interface=org.freedesktop.DBus
+      member=GetConnectionUnixUser
+      peer=(label=unconfined),
+
+  # Allow binding the service to the requested connection name
+  dbus (bind)
+      bus=system
+      name="org.freedesktop.fwupd",
+`)
+
+var fwupdConnectedPlugAppArmor = []byte(`
+# Description: Allow using fwupd service. Reserved because this gives
+#  privileged access to the fwupd service.
+# Usage: reserved
+
+  # DBus accesses
+  #include <abstractions/dbus-strict>
+
+  # Allow access to fwupd service
+  dbus (receive, send)
+      bus=system
+      path=/
+      interface=org.freedesktop.fwupd
+      peer=(label=###SLOT_SECURITY_TAGS###),
+
+  dbus (receive, send)
+      bus=system
+      path=/
+      interface=org.freedesktop.DBus.Properties
+      peer=(label=###SLOT_SECURITY_TAGS###),
+`)
+
+var fwupdConnectedSlotAppArmor = []byte(`
+# Description: Allow firmware update from fwupd service. Reserved because this gives
+#  privileged access to the system.
+# Usage: reserved
+
+  # Allow libfwup to access efivarfs with immutable flag
+  capability linux_immutable,
+
+  # File accesses
+  # Allow write access for efivarfs filesystem
+  /sys/firmware/efi/efivars/** rw,
+
+  # Allow write access for efi firmware updater
+  /boot/efi/EFI/ubuntu/fw/** rw,
+  # Allow access from efivar library
+  @{PROC}/@{pid}/mounts r,
+  /sys/devices/**/partition r,
+  /dev/sd[a-z]* r,
+  # Allow access UEFI firmware platform size
+  /sys/firmware/efi/ r,
+  /sys/firmware/efi/fw_platform_size r,
+
+  # Allow traffic to/from org.freedesktop.DBus for fwupd service
+  dbus (receive, send)
+      bus=system
+      path=/
+      interface=org.freedesktop.DBus.**
+      peer=(label=###PLUG_SECURITY_TAGS###),
+
+  dbus (receive, send)
+      bus=system
+      path=/org/freedesktop/fwupd{,/**}
+      interface=org.freedesktop.DBus.**
+      peer=(label=###PLUG_SECURITY_TAGS###),
+
+  # Allow traffic to/from fwupd interface with any method
+  dbus (receive, send)
+      bus=system
+      path=/
+      interface=org.freedesktop.fwupd
+      peer=(label=###PLUG_SECURITY_TAGS###),
+
+  dbus (receive, send)
+      bus=system
+      path=/org/freedesktop/fwupd{,/**}
+      interface=org.freedesktop.fwupd
+      peer=(label=###PLUG_SECURITY_TAGS###),
+`)
+
+var fwupdPermanentSlotDBus = []byte(`
+<policy user="root">
+    <allow own="org.freedesktop.fwupd"/>
+</policy>
+<policy context="default">
+    <allow send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.fwupd"/>
+    <allow send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.DBus.Properties"/>
+    <allow send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.DBus.Introspectable"/>
+    <allow send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.DBus.Peer"/>
+</policy>
+`)
```

Comment:
Actually, location-control may not be the perfect example. It is actually meant to be used by non-root users, but fwupd is clearly not. I'm going to amend my previous comment and say, please change the end result of the dbus bus policy (after connections are made) to be:
```
<policy user="root">
    <allow own="org.freedesktop.fwupd"/>
    <allow send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.fwupd"/>
    <allow send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.DBus.Properties"/>
    <allow send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.DBus.Introspectable"/>
    <allow send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.DBus.Peer"/>
</policy>
<policy context="default">
    <deny own="org.freedesktop.fwupd"/>
    <deny send_destination="org.freedesktop.fwupd" send_interface="org.freedesktop.fwupd"/>
</policy>
```

As it stands, when fwupd is installed, any non-snap process of any UID can talk to udisks which would provide easy privilege escalation. On a classic system we might use policykit to control this access, but all snaps does not have policy kit. If there are APIs that are meant to be accessible to everyone and are no-risk, feel free to list them in the default policy.

Note that leaving out the apparmor rule for unconfined I mentioned elsewhere does not actually prevent the privilege escalation since unconfined users can change_profile (eg, `aa-exec -p snap.fwupmgr.fwupmgr -- <something nasty>`).